editor/plugins/gradient_texture_2d_editor_plugin.cpp

```diff
@@ -0,0 +1,283 @@
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "gradient_texture_2d_editor_plugin.h"
+
+#include "editor/editor_node.h"
+#include "editor/editor_scale.h"
+#include "scene/gui/box_container.h"
+#include "scene/gui/flow_container.h"
+#include "scene/gui/separator.h"
+
+Point2 GradientTexture2DEditorRect::_get_handle_position(Handle p_handle) {
+	// Get the handle's mouse position in pixels relative to offset.
+	return (p_handle == HANDLE_FILL_FROM ? texture.ptr()->get_fill_from() : texture.ptr()->get_fill_to()) * size;
+}
+
+void GradientTexture2DEditorRect::_update_fill_position() {
+	// Update the texture's fill_from/fill_to property based on mouse input.
+	if (handle != HANDLE_NONE) {
+		Vector2 percent = (get_local_mouse_position() - offset) / size;
+		percent = Vector2(CLAMP(percent.x, 0, 1), CLAMP(percent.y, 0, 1));
+		if (snap_enabled) {
+			percent = Vector2(Math::snapped(percent.x - 0.5, snap_size) + 0.5, Math::snapped(percent.y - 0.5, snap_size) + 0.5);
+		}
+
+		String property_name = handle == HANDLE_FILL_FROM ? "fill_from" : "fill_to";
+
+		undo_redo->create_action(vformat(TTR("Set %s"), property_name), UndoRedo::MERGE_ENDS);
+		undo_redo->add_do_property(texture.ptr(), property_name, percent);
+		undo_redo->add_undo_property(texture.ptr(), property_name, handle == HANDLE_FILL_FROM ? texture.ptr()->get_fill_from() : texture.ptr()->get_fill_to());
+		undo_redo->commit_action();
+	}
+}
+
+void GradientTexture2DEditorRect::gui_input(const Ref<InputEvent> &p_event) {
+	// Grab/release handle.
+	Ref<InputEventMouseButton> mb = p_event;
+	if (mb.is_valid() && mb->get_button_index() == MouseButton::LEFT) {
+		if (mb->is_pressed()) {
+			Point2 mouse_position = mb->get_position() - offset;
+			if (Rect2(_get_handle_position(HANDLE_FILL_FROM).round() - handle_size / 2, handle_size).has_point(mouse_position)) {
+				handle = HANDLE_FILL_FROM;
+			} else if (Rect2(_get_handle_position(HANDLE_FILL_TO).round() - handle_size / 2, handle_size).has_point(mouse_position)) {
+				handle = HANDLE_FILL_TO;
+			} else {
+				handle = HANDLE_NONE;
+			}
+		} else {
+			_update_fill_position();
+			handle = HANDLE_NONE;
+		}
+	}
+
+	// Move handle.
+	Ref<InputEventMouseMotion> mm = p_event;
+	if (mm.is_valid()) {
+		_update_fill_position();
+	}
+}
+
+void GradientTexture2DEditorRect::set_texture(Ref<GradientTexture2D> &p_texture) {
+	texture = p_texture;
+	texture.ptr()->connect("changed", callable_mp((CanvasItem *)this, &CanvasItem::update));
```

Comment:
What's the purpose of using `ptr()` here?

Reply:
I didn't understand when `.ptr()` should and shouldn't be used at the time I wrote that code. I was getting errors in some places for not using it, so I added it everywhere to get it working, and then forgot to remove it in some places. I removed all the unnecessary `.ptr()` calls now.